File: frontend/lib/types/application.ts
/**
 * Application types matching backend API
 */

export interface Application {
  id: number;
  status: string;
  requirement?: string;
  work_setting: string;
  compensation?: string;
  position?: string;
  job_description?: string;
  job_link?: string;
  location?: string;
  resume?: string;
  cover_letter?: string;
  entered_iwd: number;
  date_close?: string;
  company_id?: number;
  client_id?: number;
  // Related entity names (from backend joins)
  company_name?: string;
  client_name?: string;
  contact_name?: string;
  // Related entities (from detail endpoint)
  contacts?: Array<{
    id: number;
    name: string;
    title?: string;
    contact_type: string;
    emails?: Array<{ email: string; email_type: string; is_primary: number }>;
    phones?: Array<{ phone: string; phone_type: string; is_primary: number }>;
  }>;
  company_address?: string;
  company_city?: string;
  company_state?: string;
  company_zip?: string;
  company_country?: string;
  is_deleted: number;
  created_on: string;
  modified_on: string;
  created_by: string;
  modified_by: string;
}

export interface ApplicationCreate {
  status: string;
  requirement?: string;
  work_setting: string;
  compensation?: string;
  position?: string;
  job_description?: string;
  job_link?: string;
  location?: string;
  resume?: string;
  cover_letter?: string;
  entered_iwd?: number;
  date_close?: string;
  company_id?: number;
  client_id?: number;
  created_by: string;
  modified_by: string;
}

export interface ApplicationUpdate {
  status?: string;
  requirement?: string;
  work_setting?: string;
  compensation?: string;
  position?: string;
  job_description?: string;
  job_link?: string;
  location?: string;
  resume?: string;
  cover_letter?: string;
  entered_iwd?: number;
  date_close?: string;
  company_id?: number;
  client_id?: number;
  modified_by: string;
}

File: backend/app/models/application.py
"""
Pydantic models for Application entity.
"""
from pydantic import BaseModel, Field
from typing import Optional
from datetime import datetime


class ApplicationBase(BaseModel):
    """Base model with common application fields."""
    status: str = Field(default="Pending", description="Application status")
    requirement: Optional[str] = Field(None, description="Job requirements")
    work_setting: str = Field(default="Remote", description="Work setting (Remote, Hybrid, On-site)")
    compensation: Optional[str] = Field(None, description="Compensation details")
    position: Optional[str] = Field(None, description="Job position/title")
    job_description: Optional[str] = Field(None, description="Job description")
    job_link: Optional[str] = Field(None, description="Link to job posting")
    location: Optional[str] = Field(None, description="Job location")
    resume: Optional[str] = Field(None, description="Resume used")
    cover_letter: Optional[str] = Field(None, description="Cover letter used")
    entered_iwd: int = Field(default=0, description="Entered IWD flag (0 or 1)")
    date_close: Optional[str] = Field(None, description="Date closed")
    company_id: Optional[int] = Field(None, description="Company ID")
    client_id: Optional[int] = Field(None, description="Client ID")


class ApplicationCreate(ApplicationBase):
    """Model for creating a new application."""
    created_by: str = Field(..., description="User who created the record")
    modified_by: str = Field(..., description="User who last modified the record")


class ApplicationUpdate(BaseModel):
    """Model for updating an application."""
    status: Optional[str] = None
    requirement: Optional[str] = None
    work_setting: Optional[str] = None
    compensation: Optional[str] = None
    position: Optional[str] = None
    job_description: Optional[str] = None
    job_link: Optional[str] = None
    location: Optional[str] = None
    resume: Optional[str] = None
    cover_letter: Optional[str] = None
    entered_iwd: Optional[int] = None
    date_close: Optional[str] = None
    company_id: Optional[int] = None
    client_id: Optional[int] = None
    modified_by: str = Field(..., description="User who is modifying the record")


class ApplicationResponse(ApplicationBase):
    """Model for application response."""
    id: int
    is_deleted: int = Field(default=0, description="Soft delete flag")
    created_on: datetime
    modified_on: datetime
    created_by: str
    modified_by: str

    class Config:
        from_attributes = True
